### PR TITLE
Hide experimental settings

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

Settings for experimental features should not be accessible without that experimental feature enabled. For example, if `nix-command` is enabled, but `flakes` is disabled, the CLI should refuse to:

 - accept flags which would modify flake-specific settings
 - list flake-specific settings in `help` and `show-config`
 - etc.

This commit allows marking settings experimental, like we do for built-ins already, in a declarative way.
It also allows flags to likewise be marked experimental, which is needed to implement the above but also useful in its own right.

Even if we were to stabilize `flakes` and `nix-command` at once (and I hope not!), it is still nice to have this functionality for future experimental features. For example, we might end up needing settings and flags that exist for `ca-derivations`. We should likewise be able to hide those from the stable CLI.

# Context

In #7608 @roberth started investigating what would be needed to make a very simple command `nix store gc`, Hiding these `flake` settings was one of the items in his check list.

The first commit message as a test that fails, and the latter two fix the test, until it passes. The latter two have somewhat in-depth commit messages explaining the implementation strategy.

Needed for #7608.

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [ ] agreed on idea
 - [ ] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or bug fix: updated release notes

### CODE REVIEW

File: src/libutil/config.cc [hunk 1]
```diff
@@ -70,10 +70,17 @@ void AbstractConfig::reapplyUnknownSettings()
         set(s.first, s.second);
 }
 
+static bool regular(const Config::SettingData & sd)
```
Comment (roberth) [inline src/libutil/config.cc:73]:
```suggestion
// Whether we should process the option. Excludes aliases, which are handled elsewhere, and disabled features.
static bool applicable(const Config::SettingData & sd)
```

I think this name is a bit more meaningful. Where are aliases handled?

File: src/libutil/config.hh [hunk 2]
```diff
@@ -353,4 +359,23 @@ struct GlobalConfig : public AbstractConfig
...
+
+    /* null ptr means no feature, which means there is nothing that could be
+     * disabled, and so the function returns true.
+     */
+    bool isEnabled(const ExperimentalFeature *) const;
+
+    void require(const ExperimentalFeature &) const;
+};
+
+// FIXME: don't use a global variable.
```
Comment (roberth) [inline src/libutil/config.hh:378]:
```suggestion
```

I agree, but I don't think this is an isolated problem.

File: tests/experimental-features.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+# Simple case, the configuration effects the running command
+both_ways show-config
+
+# Complicated case, earlier args effect later args
+
+both_ways store gc --help
+
+! nix --experimental-features 'nix-command' show-config --flake-registry 'https://no'
+nix --experimental-features 'nix-command flakes' show-config --flake-registry 'https://no'
```
Comment (roberth) [inline tests/experimental-features.sh:19]: A unit test would be more future proof, but this will do.

File: src/libutil/config.hh [hunk 4]
```diff
@@ -353,4 +359,23 @@ struct GlobalConfig : public AbstractConfig
...
 
+
+struct ExperimentalFeatureSettings : Config {
+
+    Setting<std::set<ExperimentalFeature>> experimentalFeatures{this, {}, "experimental-features",
+        "Experimental Nix features to enable."};
+
+    bool isEnabled(const ExperimentalFeature &) const;
+
+    /* null ptr means no feature, which means there is nothing that could be
```
Comment (roberth) [inline src/libutil/config.hh:370]:
```suggestion
    /* nullptr means no feature, which means there is nothing that could be
```

or _null pointer_

**Review by roberth — COMMENTED** [review]
Looks decent; just a suggestion and two nitpicks.
I don't think I know enough about how the initialization order of the settings to judge that aspect, but otherwise lgtm.

File: src/libutil/config.hh (hunk 4 above)
Comment (Ericson2314) [reply src/libutil/config.hh:370]:
```suggestion
    /* null pointer means no feature, which means there is nothing that could be
```

File: src/libutil/config.cc (hunk 1 above)
Comment (Ericson2314) [reply src/libutil/config.cc:73]: Not sure, but I didn't that.

File: src/libutil/config.cc [hunk 5]
```diff
@@ -147,34 +154,38 @@ nlohmann::json Config::toJSON()
 {
     auto res = nlohmann::json::object();
     for (auto & s : _settings)
-        if (!s.second.isAlias) {
+        if (regular(s.second))
```
Comment (Ericson2314) [inline src/libutil/config.cc:157]:
```suggestion
        if (applicable(s.second))
```

File: src/libutil/config.cc [hunk 6]
```diff
@@ -147,34 +154,38 @@ nlohmann::json Config::toJSON()
...
-        }
     return res;
 }
 
 std::string Config::toKeyValue()
 {
     auto res = std::string();
     for (auto & s : _settings)
-        if (!s.second.isAlias) {
+        if (regular(s.second))
```
Comment (Ericson2314) [inline src/libutil/config.cc:166]:
```suggestion
        if (applicable(s.second))
```

File: src/libutil/config.hh (hunk 2 above)
Comment (Ericson2314) [reply src/libutil/config.hh:378]: It isn't, but another such global already had this FIXME, so I just copied it for consistency.

File: tests/experimental-features.sh (hunk 3 above)
Comment (Ericson2314) [reply tests/experimental-features.sh:19]: Thanks. Yeah I would have to mock up a whole command thingy and parse it, at which point I'd be worried the mocking and the real world got out of sync.

File: src/libutil/config.hh (hunk 2 above)
Comment (roberth) [reply src/libutil/config.hh:378]: I must have made a mistake grepping for it. sgtm

File: src/libutil/args.cc [hunk 7]
```diff
@@ -124,10 +124,21 @@ void Args::parseCmdline(const Strings & _cmdline)
 
 bool Args::processFlag(Strings::iterator & pos, Strings::iterator end)
 {
+    auto expFlag = [](const Args::Flag & f) -> bool {
+        auto & xp = f.experimentalFeature;
+        return experimentalFeatureSettings.isEnabled(xp ? &*xp : nullptr);
+    };
+
     assert(pos != end);
```
Comment (roberth) [inline src/libutil/args.cc:132]: I'm getting
```
$ nix shell ~/h/nix --accept-flake-config
nix: src/libutil/args.cc:132: virtual bool nix::Args::processFlag(std::__cxx11::list<std::__cxx11::basic_string<char> >::iterator&, std::__cxx11::list<std::__cxx11::basic_string<char> >::iterator): Assertion `pos != end' failed.
Aborted (core dumped)
```

File: tests/experimental-features.sh [hunk 8]
```diff
@@ -0,0 +1,19 @@
+source common.sh
+
+# Without flakes, flake options should not show up
+# With flakes, flake options should show up
+
+function both_ways {
+    nix --experimental-features 'nix-command' "$@" | grepQuietInverse flake
+    nix --experimental-features 'nix-command flakes' "$@" | grepQuiet flake
```
Comment (roberth) [inline tests/experimental-features.sh:8]: Flags used to be declarative. This currently breaks:

```suggestion
    nix --experimental-features 'nix-command' "$@" | grepQuietInverse flake
    nix --experimental-features 'nix-command flakes' "$@" | grepQuiet flake

    # Also, the order should not matter
    nix "$@" --experimental-features 'nix-command' | grepQuietInverse flake
    nix "$@" --experimental-features 'nix-command flakes'  | grepQuiet flake
```

File: src/libutil/config.cc [hunk 9]
```diff
@@ -444,4 +462,30 @@ GlobalConfig::Register::Register(Config * config)
...
+
+bool ExperimentalFeatureSettings::isEnabled(const ExperimentalFeature * feature) const
+{
+    return !feature || isEnabled(*feature);
+}
+
+void ExperimentalFeatureSettings::require(const ExperimentalFeature * feature) const
+{
+    if (feature) require(*feature);
+}
```
Comment (roberth) [inline src/libutil/config.cc:489]: Would be nicer to use `std::optional<>&`, but otherwise the pr lgtm!

**Comment by fricklerhandwerk** [comment]
Discussed in the Nix team meeting 2023-03-13:

 - Annoying because of the churn coming with it
 - But also very useful as long as we have experimental features.
 - Idea approved, assigned to @roberth

**Pull request merged**